### ISSUE TITLE
Remove menu-cache

### ISSUE THREAD

**Issue by tsujan** [lxqt/libfm-qt#417]
##### Expected Behavior

No menu-cache dependency.
##### Current Behavior

libfm-qt/pcmanfm-qt still depends on menu-cache for some functionalities. menu-cache has shown problematic behavior before (long delays in reboot/shutdown) and depends on libfm-extra
##### Possible Solution

I have no idea yet.

##### Context

Please see https://github.com/lxqt/pcmanfm-qt/issues/952 and https://github.com/lxqt/libfm-qt/issues/185.

IMHO, menu-cache shouldn't be made optional but removed.
##### System Information

Latest git LXQt

**Comment by tsujan** [comment]
Sadly, nothing has been done after 4 years. IMO, using `libqtxdg` instead of `menu-cache` isn't the solution. `libfm-qt` needs to do all the job internally, without depending on a separate library. That requires lots of changes, for which I haven't find the time.

**Comment by isf63** [comment]
Is the menu-cache daemon still in use? I don't see `menu-cached` in `ps -A`...

I can compile ./libmenu-cache/menu-cache.c with `g++ menu-cache.c `pkg-config --cflags --libs glib-2.0 gio-2.0` -fpermissive -DMENUCACHE_LIBEXECDIR=""` with two files, menu-cache.h.in renamed to menu-cache.h, and version.h.

If the daemon is not used then a simple copy/translating of those three files might work.

**Comment by tsujan** [comment]
> Is the menu-cache daemon still in use?

Yes. I haven't done anything about it. menu-cache is used by Applications.

**Comment by tsujan** [comment]
Since I opened this issue, I started to have doubts. The more I think about it, the more menu-cache seems unavoidable to me. The situation is quite different from that of lxqt-panel or lxqt-runner:

In libfm-qt, we need a tool to make a hierarchy of apps, both for handling file associations and for `menu://applications/`. That's possible by using purely Qt methods and a huge amount of new codes. However, since we use GLib for file management, that tool should also be based on GLib. I can think of nothing better than menu-cache, which extends GLib for this job.

It's possible to internalize it, but that's like reinventing the wheel. menu-cache seems to be maintained well; if at some point it isn't, we might want to adapt its codes inside libfm-qt.

**Comment by tsujan** [comment]
@ahsand97 
Can you think of a "magic" solution? I don't suggest wasting your time on this while you could do better things. Please just read my last comment and take a look at the code.

**Comment by ahsand97** [comment]
> Can you think of a "magic" solution?

Hmm this one is hard to be solved with little work but...I can think of 2 solutions and one of them is really really not nice because it would require reinvent the wheel as you said.

**Option A — Scoped parser matching the current `lxqt-applications.menu`**

Parse `lxqt-applications.menu` (and `applications-merged/*` per `<DefaultMergeDirs/>`) directly inside libfm-qt using `QXmlStreamReader`, building an in-memory tree of `<Menu>` nodes (name, `.directory` reference, include/exclude rule, `OnlyUnallocated` flag, child menus).

- A small recursive rule evaluator supports exactly the tags the file uses today: `Category`, `Filename`, `And`, `Or`, `Not`.
- Installed apps are enumerated via GIO's `g_app_info_get_all()` + `g_app_info_should_show()` (which already implements `Hidden`/`NoDisplay`/`OnlyShowIn`/`TryExec` filtering correctly), then matched against the rule tree, with a "claimed" set tracked during the walk so `OnlyUnallocated` (the `Other` menu) resolves correctly.
- Each menu node's display name/icon comes from its `.directory` file via `GKeyFile`.
- Any tag outside this supported subset (e.g. an explicit-path `<MergeFile>`, `<LegacyDir>`) logs a warning instead of silently no-opping, and a CI test runs the parser against the shipped menu file so an unsupported construct fails the build rather than shipping broken. (I don't see any CI workflow in the project so I don't know if this would be possible).
- Reload on change via a file watcher on the menu file and merge directories.
- Also fixes `appchooserdialog.cpp`'s small `menu_cache_lookup()` call by widening its existing `g_app_info_get_all_for_type()` search to `g_app_info_get_all()`.

Scope is bounded to the spec constructs `lxqt-applications.menu` currently uses; broader spec support could be added incrementally later if a menu-file change ever needs it.


This one seems very reasonable to me and doable, and it wouldn't be as big as rewriting everything that `menu-cache` does, like implementing the whole freedesktop.org Desktop Menu Specification.

---

**Option B — Full independent Desktop Menu Specification implementation**

Make libfm-qt completely implement https://specifications.freedesktop.org/menu/latest/
Same overall pipeline. But just by looking at [`menu-cache`](https://github.com/lxde/menu-cache) code it'll be like ~2900 lines of code, such a nice PR, would you imagine? haha.

**Comment by tsujan** [comment]
> Option A — Scoped parser matching the current lxqt-applications.menu

I have problem in translating codes into/from English (or my native language), perhaps because I'm addicted to visualization ;)

> this one is hard to be solved with little work.

Yes, I know, but I thought you might have a fresh perspective on it — which apparently you do :)

Take your time and feel free to do it in whatever way you see fit, but *only if you enjoy it*; I absolutely don't want to impose it on you as a task. If it isn't interesting to you, just ignore my request; I'll fully understand.

It was and is interesting to me but sadly I haven't found the time. Taking care of my own programs as well as some LXQt ones, I'm sometimes bombarded with new problems.

**Comment by ahsand97** [comment]
@tsujan I can do it and yeah I do enjoy it :), getting familiar with lxqt's different components has been really fun to me hehe, specially since I'm also a daily user. The thing I want to make clear is that the PR will be big, it may take me some days but if we're willing to review and have this implemented I can definitely give it a try!.

**Comment by tsujan** [comment]
>  the PR will be big, it may take me some days

There's no haste. The removal of menu-cache and making direct use of GLib functions will be a big step which will need a big PR, a detailed review, and thorough tests. It might take weeks or months. We don't expect that it'll be ready for the next release.

**Comment by tsujan** [comment]
@yan12125 
As far as I know, `menu-cache` won't be needed with LXQt ≥ 2.5.0.

**Comment by yan12125** [comment]
Cool!
